3ds2/src/main/java/com/adyen/checkout/adyen3ds2/internal/ui/DefaultAdyen3DS2Delegate.kt

```diff
@@ -404,32 +400,59 @@ internal class DefaultAdyen3DS2Delegate(
         }
     }
 
-    override fun cancelled() {
+    private fun onCancelled(result: ChallengeResult.Cancelled) {
         Logger.d(TAG, "challenge cancelled")
+
+        try {
+            val details = makeDetails(result.transactionStatus, result.additionalDetails)
+            emitDetails(details)
+        } catch (e: CheckoutException) {
+            exceptionChannel.trySend(e)
+        } finally {
+            closeTransaction()
+        }
+
         exceptionChannel.trySend(Cancelled3DS2Exception("Challenge canceled."))
-        closeTransaction()
     }
 
-    override fun timedout() {
+    private fun onTimeout(result: ChallengeResult.Timeout) {
         Logger.d(TAG, "challenge timed out")
+        try {
+            val details = makeDetails(result.transactionStatus, result.additionalDetails)
+            emitDetails(details)
+        } catch (e: CheckoutException) {
+            exceptionChannel.trySend(e)
+        } finally {
+            closeTransaction()
+        }
         exceptionChannel.trySend(Authentication3DS2Exception("Challenge timed out."))
         closeTransaction()
     }
 
-    override fun protocolError(protocolErrorEvent: ProtocolErrorEvent) {
-        with(protocolErrorEvent.errorMessage) {
-            Logger.e(TAG, "protocolError - $errorCode - $errorDescription - $errorDetails")
-            exceptionChannel.trySend(Authentication3DS2Exception("Protocol Error - $this"))
+    private fun onError(result: ChallengeResult.Error) {
+        Logger.d(TAG, "challenge timed out")
+        try {
+            val details = makeDetails(result.transactionStatus, result.additionalDetails)
+            emitDetails(details)
+        } catch (e: CheckoutException) {
+            exceptionChannel.trySend(e)
+        } finally {
+            closeTransaction()
         }
+        exceptionChannel.trySend(Authentication3DS2Exception("An error occured during challenge."))
         closeTransaction()
```

Comment:
`closeTransaction()` will be called twice in this flow right?

Reply:
Removed the extra call.